### PR TITLE
remove single-argument methods for `map`, `foreach`, `Iterators.map`

### LINKED ISSUE #35293: [RFC] Remove map(f) and foreach(f)
It might be useful to define curried version of iterator transforms such as `Iterators.map(f) = x -> Iterators.map(f, x)` (after #34352) and `Iterators.filter(f) = x -> Iterators.filter(f, x)` (ref #33027, https://github.com/JuliaLang/julia/pull/24990#issuecomment-605384271).  However, since the eager counter part `map(f)` is currently defined as `f()`, it may confuse users when the difference between `map(f)` and `Iterators.map(f)` are more than laziness.

Can we remove `map(f)` at some point? (I think whether or not it is desirable to do this within Julia 1.x is up to core devs.)  I think it makes sense to also remove `foreach(f)` when removing `map(f)` for consistency.

`map(f)` was added by @stevengj in #17318. The reasoning was that it is better to be consistent with `f.()`. I agree that considering `f.()` as a special case of `f.(args...)` where `args` are numbers make sense for broadcasting; i.e., the output dimension of broadcasting is `reduce(max, ndims_of_arguments)` when `ndims_of_arguments` is not empty.  Since `ndims` is always a non-negative integer, it is reasonable to use the identity element of `max` on non-negative integers which is `0`, when there is no argument.  However, this argument relies on that the "`ndims`" of the output is flexibly determined (i.e., using the largest `ndims`) in broadcasting.  As this is not the case in `map`, I think it makes less sense to define `map(f)` as `f()`.  OTOH, I think curried `map` is a useful definition.

`foreach(f)` was added by @JeffBezanson in #13774.  It seems there was no discussion for adding `foreach(f)`.

### PR DESCRIPTION
**Pull request by nsajko**
In the case of `map` and `foreach`, this removes awkward functionality. `Iterators.map(::Any)`, on the other hand, already throwed anyway.

Fixes #35293

### CODE REVIEW

**Comment by nsajko** [comment]
The test failures seem unrelated (they're system-specific).

**Comment by nsajko** [comment]
I guess this should get a Nanosoldier PkgEval run.

**Comment by nsajko** [comment]
Fixed conflict.

**Comment by maleadt** [comment]
@nanosoldier `runtests()`

**Comment by andyferris** [comment]
I, for one, support this change. LGTM.

**Comment by timholy** [comment]
@nsajko if you can analyze the errors and report whether they're real, that would likely help decide whether this is feasible.

**Comment by nsajko** [comment]
@timholy the analysis is on the #35293 page. One of your packages, ProgressMeter, is mentioned.

File: NEWS.md
Comment (LilithHafner) [inline NEWS.md:174]:
```suggestion
* `Base.map`, `Iterators.map`, and `foreach` lost their single-argument methods ([#52631]).
```

**Review by LilithHafner — APPROVED** [review]
Given widespread approval, limited opposition, triage approval, and a nanosoldier run that, according to @nsajko's [detailed analysis](https://github.com/JuliaLang/julia/issues/35293#issuecomment-1897756235), revealed no real breakages, this LGTM from a breakage standpoint.

The implementation looks complete and sufficiently tested.

**Comment by mikmoore** [comment]
Should `reduce`/`foldl`/`foldr`, and their `map*` variants have been included with this change? I'm sure there are arguments for either direction.

**Comment by LilithHafner** [comment]
I think they should have also been included. If someone makes a PR soon I think those removals should be backported to 1.11 so they land with this one.

As long as they are not widely used, I oppose the existence of those methods on the same grounds as `map(f) = [f()]`: the mathematically consistent implementation is `map(f) = [f() for _ in 1:Inf]`, which folks don't want.

**Comment by nsajko** [comment]
> Should `reduce`/`foldl`/`foldr`, and their `map*` variants have been included with this change?

Those weren't part of the linked issue, so presumably they're not covered with the triage decision. A new issue should be opened IMO, and discussed by triage I guess.

**Comment by LilithHafner** [comment]
No need to get triage approval to open a PR! (plus, we need a PR to run nanosoldier which will inform decisionmaking)

**Comment by LilithHafner** [comment]
(Though I would want to bring this back to triage before merging such a PR)

**Comment by dlfivefifty** [comment]
> `[f() for _ in 1:Inf]`

In case this is something you actually want this works:
```julia
julia> using InfiniteArrays

julia> f = () -> 1
#5 (generic function with 1 method)

julia> [f() for _ = 1:∞]
#7.(ℵ₀-element InfiniteArrays.InfUnitRange{Int64} with indices OneToInf()) with indices OneToInf():
 1
 1
 1
 1
 1
 1
 1
 1
 1
 ⋮
```

**Comment by nsajko** [comment]
Followup: #54088

**Pull request merged**
